File: modal-go/test/auth_token_manager_test.go
package test

import (
	"context"
	"testing"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/modal-labs/libmodal/modal-go"
	pb "github.com/modal-labs/libmodal/modal-go/proto/modal_proto"
	"github.com/onsi/gomega"
	"google.golang.org/grpc"
)

type mockAuthClient struct {
	pb.ModalClientClient
	authToken string
}

func newMockAuthClient() *mockAuthClient {
	return &mockAuthClient{}
}

func (m *mockAuthClient) setAuthToken(token string) {
	m.authToken = token
}

func (m *mockAuthClient) AuthTokenGet(ctx context.Context, req *pb.AuthTokenGetRequest, opts ...grpc.CallOption) (*pb.AuthTokenGetResponse, error) {
	return pb.AuthTokenGetResponse_builder{
		Token: m.authToken,
	}.Build(), nil
}

// Creates a JWT token for testing
func createTestJWT(expiry int64) string {
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
		"exp": expiry,
		"iat": time.Now().Unix(),
	})

	tokenString, _ := token.SignedString([]byte("walter-test"))
	return tokenString
}

func TestAuthTokenManager_DecodeJWT(t *testing.T) {
	g := gomega.NewWithT(t)

	mockClient := newMockAuthClient()
	manager := modal.NewAuthTokenManager(mockClient)

	validToken := createTestJWT(123456789)
	mockClient.setAuthToken(validToken)

	// FetchToken should decode and store the JWT
	_, err := manager.FetchToken(context.Background())
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	g.Expect(manager.GetCurrentToken()).Should(gomega.Equal(validToken))
}

// Setting the initial token and having it cached.
func TestAuthTokenManager_InitialFetch(t *testing.T) {
	g := gomega.NewWithT(t)

	mockClient := newMockAuthClient()
	token := createTestJWT(time.Now().Unix() + 3600)
	mockClient.setAuthToken(token)

	manager := modal.NewAuthTokenManager(mockClient)
	err := manager.Start(context.Background())
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	defer manager.Stop()

	firstToken, firstErr := manager.GetToken(context.Background())
	g.Expect(firstErr).ShouldNot(gomega.HaveOccurred())
	g.Expect(firstToken).Should(gomega.Equal(token))

	secondToken, secondErr := manager.GetToken(context.Background())
	g.Expect(secondErr).ShouldNot(gomega.HaveOccurred())
	g.Expect(secondToken).Should(gomega.Equal(token))
}

func TestAuthTokenManager_IsExpired(t *testing.T) {
	g := gomega.NewWithT(t)

	manager := modal.NewAuthTokenManager(nil)

	// Test not expired
	manager.SetToken("token", time.Now().Unix()+3600)
	g.Expect(manager.IsExpired()).Should(gomega.BeFalse())

	// Test expired
	manager.SetToken("token", time.Now().Unix()-3600)
	g.Expect(manager.IsExpired()).Should(gomega.BeTrue())
}

// Refreshing an expired token. Unlikely to occur since we refresh in background before expiry.
func TestAuthTokenManager_RefreshExpiredToken(t *testing.T) {
	g := gomega.NewWithT(t)

	mockClient := newMockAuthClient()
	now := time.Now().Unix()

	expiringToken := createTestJWT(now - 60)
	freshToken := createTestJWT(now + 3600)

	manager := modal.NewAuthTokenManager(mockClient)
	manager.SetToken(expiringToken, now-60)
	mockClient.setAuthToken(freshToken)

	// Start the background refresh goroutine
	err := manager.Start(context.Background())
	g.Expect(err).ToNot(gomega.HaveOccurred())
	defer manager.Stop()

	// Wait for background goroutine to refresh the token
	g.Eventually(func() string {
		return manager.GetCurrentToken()
	}, "1s", "10ms").Should(gomega.Equal(freshToken))
}

func TestAuthTokenManager_RefreshNearExpiryToken(t *testing.T) {
	g := gomega.NewWithT(t)

	mockClient := newMockAuthClient()
	now := time.Now().Unix()

	expiringToken := createTestJWT(now + 60)
	freshToken := createTestJWT(now + 3600)

	manager := modal.NewAuthTokenManager(mockClient)
	manager.SetToken(expiringToken, now+60)
	mockClient.setAuthToken(freshToken)

	// Start the background refresh goroutine
	err := manager.Start(context.Background())
	g.Expect(err).ToNot(gomega.HaveOccurred())
	defer manager.Stop()

	// Wait for background goroutine to refresh the token
	g.Eventually(func() string {
		return manager.GetCurrentToken()
	}, "1s", "10ms").Should(gomega.Equal(freshToken))
}

// Should error out if no valid token is available.
func TestAuthTokenManager_GetToken_ExpiredToken(t *testing.T) {
	g := gomega.NewWithT(t)

	mockClient := newMockAuthClient()
	manager := modal.NewAuthTokenManager(mockClient)

	_, err := manager.GetToken(context.Background())
	g.Expect(err).Should(gomega.HaveOccurred())
}

func TestClient_Close(t *testing.T) {
	g := gomega.NewWithT(t)

	mockClient := newMockAuthClient()
	token := createTestJWT(time.Now().Unix() + 3600)
	mockClient.setAuthToken(token)

	client, err := modal.NewClientWithOptions(&modal.ClientParams{
		ControlPlaneClient: mockClient,
	})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	client.Close()
}

func TestClient_MultipleInstancesSeparateManagers(t *testing.T) {
	g := gomega.NewWithT(t)

	mockClient1 := newMockAuthClient()
	token1 := createTestJWT(time.Now().Unix() + 3600)
	mockClient1.setAuthToken(token1)

	mockClient2 := newMockAuthClient()
	token2 := createTestJWT(time.Now().Unix() + 3600)
	mockClient2.setAuthToken(token2)

	client1, err1 := modal.NewClientWithOptions(&modal.ClientParams{
		ControlPlaneClient: mockClient1,
	})
	g.Expect(err1).ShouldNot(gomega.HaveOccurred())
	defer client1.Close()

	client2, err2 := modal.NewClientWithOptions(&modal.ClientParams{
		ControlPlaneClient: mockClient2,
	})
	g.Expect(err2).ShouldNot(gomega.HaveOccurred())
	defer client2.Close()
}

File: modal-js/test/legacy/auth_token_manager.test.ts
import { describe, test, expect, vi, beforeEach, afterEach } from "vitest";
import jwt from "jsonwebtoken";
import { AuthTokenManager } from "../../src/auth_token_manager";

async function eventually(
  condition: () => boolean,
  timeoutMs: number = 1000,
  intervalMs: number = 10,
): Promise<void> {
  const start = Date.now();
  while (Date.now() - start < timeoutMs) {
    if (condition()) {
      return;
    }
    await new Promise((resolve) => setTimeout(resolve, intervalMs));
  }
  throw new Error(`Condition not met within ${timeoutMs}ms`);
}

class mockAuthClient {
  private authToken: string = "";

  setAuthToken(token: string) {
    this.authToken = token;
  }

  authTokenGet = vi.fn(async () => {
    return { token: this.authToken };
  });
}

function newMockAuthClient() {
  return new mockAuthClient();
}

// Creates a JWT token for testing
function createTestJWT(expiry: number): string {
  return jwt.sign({ exp: expiry }, "walter-test");
}

describe("AuthTokenManager", () => {
  let mockClient: mockAuthClient;
  let manager: AuthTokenManager;

  beforeEach(() => {
    mockClient = newMockAuthClient();
    manager = new AuthTokenManager(mockClient as any);
  });

  afterEach(() => {
    manager.stop();
  });

  test("TestAuthToken_DecodeJWT", async () => {
    const now = Math.floor(Date.now() / 1000);
    const expiry = now + 1800;
    const token = createTestJWT(expiry);
    mockClient.setAuthToken(token);

    // Test by fetching a valid JWT and checking if it gets stored properly
    await manager.start();
    expect(manager.getCurrentToken()).toBe(token);
  });

  test("TestAuthToken_InitialFetch", async () => {
    const now = Math.floor(Date.now() / 1000);
    const token = createTestJWT(now + 3600);
    mockClient.setAuthToken(token);

    await manager.start();

    const firstToken = await manager.getToken();
    expect(firstToken).toBe(token);

    const secondToken = await manager.getToken();
    expect(secondToken).toBe(token);
  });

  test("TestAuthToken_IsExpired", async () => {
    const now = Math.floor(Date.now() / 1000);

    // Test not expired
    const validToken = createTestJWT(now + 3600);
    manager.setToken(validToken, now + 3600);
    expect(manager.isExpired()).toBe(false);

    // Test expired
    const expiredToken = createTestJWT(now - 60);
    manager.setToken(expiredToken, now - 60);
    expect(manager.isExpired()).toBe(true);
  });

  test("TestAuthToken_RefreshExpiredToken", async () => {
    const now = Math.floor(Date.now() / 1000);
    const expiringToken = createTestJWT(now - 60);
    const freshToken = createTestJWT(now + 3600);

    manager.setToken(expiringToken, now - 60);
    mockClient.setAuthToken(freshToken);

    // Start the background refresh
    await manager.start();

    // Wait for background refresh to update the token
    await eventually(() => manager.getCurrentToken() === freshToken);

    // Should have the new token cached
    expect(manager.getCurrentToken()).toBe(freshToken);
  });

  test("TestAuthToken_RefreshNearExpiryToken", async () => {
    const now = Math.floor(Date.now() / 1000);
    const expiringToken = createTestJWT(now + 60);
    const freshToken = createTestJWT(now + 3600);

    manager.setToken(expiringToken, now + 60);
    mockClient.setAuthToken(freshToken);

    // Start the refresh timer
    await manager.start();

    // Wait for background refresh to update the token
    await eventually(() => manager.getCurrentToken() === freshToken);

    // Should have the new token cached
    expect(manager.getCurrentToken()).toBe(freshToken);
  });

  test("TestAuthToken_ConcurrentGetToken", async () => {
    const token = createTestJWT(Math.floor(Date.now() / 1000) + 3600);
    mockClient.setAuthToken(token);

    // Start the manager to fetch initial token
    await manager.start();

    // Multiple concurrent getToken calls should all return the same token
    const [result1, result2, result3] = await Promise.all([
      manager.getToken(),
      manager.getToken(),
      manager.getToken(),
    ]);
    expect(result1).toBe(token);
    expect(result2).toBe(token);
    expect(result3).toBe(token);

    // authTokenGet should have been called only once (during start)
    expect(mockClient.authTokenGet).toHaveBeenCalledTimes(1);
  });

  test("TestAuthToken_GetToken_NoToken", async () => {
    await expect(manager.getToken()).rejects.toThrow(
      "No valid auth token available",
    );
  });
});
